### PR TITLE
Add Speculative Decoding Eagle3 topk > 1

### PR DESCRIPTION
**Pull request by qingquansong**
## Motivation

Add Speculative Decoding Eagle3 topk > 1

```
GSM8k - 8 shots (with cuda graph) --speculative-num-steps 3  --speculative-eagle-topk 2 --speculative-num-draft-tokens 3 --dtype bfloat16

Accuracy: 0.792
Invalid: 0.001
Latency: 43.638 s
Output throughput: 3058.043 token/s
```

```
send one  (with cuda graph) --speculative-num-steps 3 --speculative-eagle-topk 2 --speculative-num-draft-tokens 3 --dtype bfloat16

acc_length=2.23
speed=195.98 token/s
```

### TODOs in next PR
TODO(hebiao064): Accelerate FA3 Spec Decode with topk > 1.
TODO(hebiao064): Improve the acc rate for FA3 Spec Decode with topk == 1 and page_size > 1.

#### 🧠 LLaMA 3.1 8B

| Spec      | Acc. Length | Speed (token/s) |
|-----------|-------------|-----------------|
| 548       | 2.56        | 204.46          |
| 314       | 2.12        | 214.14          |
| No Spec   | 1.00        | 148.66          |

---

#### 🧪 DS V3 (Actual Model!)

| Spec      | Acc. Length | Speed (token/s) |
|-----------|-------------|-----------------|
| 548       | 3.57        | 101.92          |
| 324       | 2.94        | 97.95           |
| 122       | 1.94        | 70.34           |
| 314       | 2.77        | 96.64           |
| No Spec   | 1.00        | 50.08           |

---

#### 🔬 DS CI Model

| Spec      | Acc. Length | Speed (token/s) |
|-----------|-------------|-----------------|
| 548       | 3.79        | 165.92          |
| 314       | 3.03        | 163.56          |
| No Spec   | 1.00        | 92.02           |

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -587,12 +746,42 @@ def forward_extend(
...
+                    max_seqlen_q=self.forward_metadata_spec_decode_expand.max_seq_len_q,
+                    softmax_scale=layer.scaling,
+                    causal=False,
+                    window_size=window_size,
+                    softcap=layer.logit_cap,
+                    k_descale=k_descale,
+                    v_descale=v_descale,
+                    return_softmax_lse=True,
+                )
+                o, _ = merge_state(
```
Comment (zcnrex) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: Do we need to check `is_flashinfer_available` here before using `merge_state`

Comment (zhyncs) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: Let’s migrate it to sgl-kernel

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -1555,31 +1601,35 @@ def init_forward_metadata_replay_cuda_graph(
...
 
-            # Update the regular page table
-            page_table = self.req_to_token[
-                req_pool_indices,
-                metadata.encoder_max_seq_len_k : (
-                    metadata.encoder_max_seq_len_k + metadata.max_seq_len_k
-                ),
-            ]
-            metadata.page_table[:, : metadata.max_seq_len_k].copy_(page_table)
+        # if encoder_lens is not None:
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: this is still needed for MM

Comment (qingquansong) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: yep yep, let me revert it. Thanks!

**Comment by hebiao064** [comment]
After merged with @Fridge003 's [PR](https://github.com/sgl-project/sglang/pull/5113) with conflict resolved, we can verify chunked attention optimization still valid:

```bash
python3 -m sglang.launch_server --model /shared/public/elr-models/deepseek-ai/DeepSeek-V3/1d044fd82b15f1cedb197a288e50cc96a2c27205/  --disable-radix-cache --max-prefill-tokens 32768  --tp 8

Backend:                                 sglang    
Traffic request rate:                    8.0       
Max reqeuest concurrency:                not set   
Successful requests:                     120       
Benchmark duration (s):                  311.36    
Total input tokens:                      3840000   
Total generated tokens:                  12000     
Total generated tokens (retokenized):    11914     
Request throughput (req/s):              0.39      
Input token throughput (tok/s):          12332.86  
Output token throughput (tok/s):         38.54     
Total token throughput (tok/s):          12371.40 
```

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 3]
```diff
@@ -267,6 +268,11 @@ def cdiv(a: int, b: int) -> int:
     return -(a // -b)
 
 
+@torch._dynamo.disable()
```
Comment (zhyncs) [inline python/sglang/srt/layers/attention/flashattention_backend.py:272]: @DefTruth

File: test/srt/test_fa3.py [hunk 4]
```diff
@@ -195,6 +195,8 @@ def get_server_args(cls):
                 "1",
                 "--speculative-num-draft-tokens",
                 "3",
+                "--dtype",
+                "float16",
```
Comment (zhyncs) [inline test/srt/test_fa3.py]: Why do we need this?

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 5]
```diff
@@ -558,13 +717,41 @@ def forward_extend(
...
+                causal=(
+                    False
+                    if forward_batch.forward_mode.is_target_verify() and self.topk > 1
+                    else causal
+                ),
                 window_size=window_size,
                 softcap=layer.logit_cap,
                 k_descale=k_descale,
                 v_descale=v_descale,
+                return_softmax_lse=True,
```
Comment (zhyncs) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: Do we always need to set return_softmax_lse to true? I think it may introduce overhead. Can we only use it for top k > 1 for all flash_attn_with_kvcache use cases?

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 3 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:272]: we need to disable torch compile to make it work

File: test/srt/test_fa3.py (hunk 4 above)
Comment (hebiao064) [reply test/srt/test_fa3.py]: accidentally added, will remove

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 3 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:272]: added a todo for this line, will address it later

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 5 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: make sense, addressed

**Comment by hebiao064** [comment]
Benchmark from recent commit:

```bash
Final Test for cuda graph + torch compile:

- Llama 323: Send one work
    - acc_length=1.84
    speed=126.15 token/s
- Llama 313: Send one work
    - acc_length=1.79
    speed=176.51 token/s
- Llama no spec
    - acc_length=1.00
    speed=144.86 token/s
- DS 323:
    - acc_length=2.60
    speed=139.08 token/s
- DS 313:
    - acc_length=2.60
    speed=145.92 token/s
- DS No Spec:
    - acc_length=1.00
    speed=94.24 token/s
```

File: python/sglang/srt/model_executor/model_runner.py [hunk 6]
```diff
@@ -227,7 +227,16 @@ def model_specific_adjustment(self):
         server_args = self.server_args
 
         if server_args.attention_backend is None:
-            # By default, use flashinfer for non-mla attention and triton for mla attention
+            """
+            We auto select the fastest attention backend according to the current offering
+            1. Models with MHA Architecture (e.g: Llama, QWen)
+                1.1 We will turn on FA3 on hopper unless user use spec decode with topk > 1 or page_size > 1.
+                1.2 In other cases, we will use flashinfer if available, otherwise use triton.
+            2. Models with MHA Architecture and using FA3
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/model_runner.py]: typo: MHA -> MLA

File: test/srt/run_suite.py
Comment (Ying1123) [inline test/srt/run_suite.py:1]: Do we have a test case for topk > 1 as in this PR?

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 7]
```diff
@@ -311,8 +320,11 @@ def __init__(
         self.page_size = model_runner.page_size
         self.use_mla = model_runner.model_config.attention_arch == AttentionArch.MLA
         self.skip_prefill = skip_prefill
-
-        self.topk = topk
+        self.topk = topk or (
+            model_runner.server_args.speculative_eagle_topk
+            if model_runner.server_args.speculative_eagle_topk is not None
+            else 0
```
Comment (Ying1123) [inline python/sglang/srt/layers/attention/flashattention_backend.py]:
```suggestion
            model_runner.server_args.speculative_eagle_topk or 0
```

**Review by Ying1123 — APPROVED** [review]

**Review by merrymercy — APPROVED** [review]
Please add a test case and we can merge this soon!

File: test/srt/run_suite.py
Comment (hebiao064) [reply test/srt/run_suite.py:1]: let me add it

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 7 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py]: resolved, thanks.

I didn't accepted your commit since I don't want to dismiss your approval :)

File: test/srt/run_suite.py
Comment (hebiao064) [reply test/srt/run_suite.py:1]: added

**Comment by hebiao064** [comment]
sanity check for 6 cases all looking good:

```
Final Send One Test for cuda graph + torch compile:

# Llama 3.1 8B
- 548: 
acc_length=2.56
speed=204.46 token/s
- 314: 
acc_length=2.12
speed=214.14 token/s
- no spec
acc_length=1.00
speed=148.66 token/s

# DS V3 (Actual Model!)
- 548 
acc_length=3.57
speed=101.92 token/s
- 324
acc_length=2.94
speed=97.95 token/s
- 122
acc_length=1.94
speed=70.34 token/s
- 314
acc_length=2.77
speed=96.64 token/s
- No Spec
acc_length=1.00
speed=50.08 token/s


# DS CI Model
- 548:
acc_length=3.79
speed=165.92 token/s
-314:
acc_length=3.03
speed=163.56 token/s
- No Spec:
acc_length=1.00
speed=92.02 token/s
```

File: python/sglang/srt/model_executor/model_runner.py [hunk 8]
```diff
@@ -234,9 +243,7 @@ def model_specific_adjustment(self):
                         "flashinfer" if is_flashinfer_available() else "triton"
                     )
             else:
-                if is_hopper_with_cuda_12_3() and is_no_spec_infer_or_topk_one(
-                    server_args
-                ):
+                if is_hopper_with_cuda_12_3():
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/model_runner.py:246]: Shall we also remove `is_no_spec_infer_or_topk_one` for non MLA use cases?

Comment (hebiao064) [reply python/sglang/srt/model_executor/model_runner.py:246]:
```
We auto select the fastest attention backend according to the current offering
1. Models with MHA Architecture (e.g: Llama, QWen)
    1.1 We will turn on FA3 on hopper unless user use spec decode with topk > 1 or page_size > 1.
    1.2 In other cases, we will use flashinfer if available, otherwise use triton.
2. Models with MLA Architecture (e.g: Deepseek)
    2.1 We will use FA3 backend on hopper.
    2.2 Otherwise, we will use triton backend.
```


After we optimized the speed of TopK > 1, we can remove it.

The reason why we enable fa3 for MLA's spec decoding topk > 1 is because: Flashinfer doesn't support top k > 1 for MLA.

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 9]
```diff
@@ -357,44 +458,101 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+                metadata_expand.page_table = non_masked_page_table.gather(1, sort_order)
+                metadata_expand.cache_seqlens_int32 = mask.sum(dim=1).to(torch.int32)
+                metadata_expand.cu_seqlens_k = torch.nn.functional.pad(
+                    torch.cumsum(
+                        metadata_expand.cache_seqlens_int32, dim=0, dtype=torch.int32
+                    ),
+                    (1, 0),
+                )
+                metadata_expand.max_seq_len_k = (
+                    metadata_expand.cache_seqlens_int32.max().item()
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashattention_backend.py:553]: an extra device sync

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 10]
```diff
@@ -357,44 +458,101 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+                cum_len = torch.nn.functional.pad(
+                    torch.cumsum(
+                        (
+                            forward_batch.seq_lens + self.speculative_num_draft_tokens
+                        ).repeat_interleave(self.speculative_num_draft_tokens),
+                        dim=0,
+                    ),
+                    (1, 0),
+                )[:-1]
+                mask_extraction_indices = (
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashattention_backend.py:510]: we can provide a better version which only generates useful custom mask

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 9 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:553]: will address in next followup pr about optimization

**Comment by Swipe4057** [comment]
Why are you running it with --dtype bfloat16? Will EAGLE's speculative decoding work when the main model is in FP8 and the draft model is in original weights? The main model in fp8 is the target use case in production.
